src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -579,26 +579,43 @@ cdef const char ** to_cstring_array(list_str):
     return ret
 
 
-def shape(font: Font, buffer: Buffer, features: Dict[str, bool] = None,
+def shape(font: Font, buffer: Buffer,
+        features: Dict[str,Union[int,bool,Sequence[Tuple[int,int,Union[int,bool]]]]] = None,
         shapers: List[str] = None) -> None:
     cdef unsigned int size
     cdef hb_feature_t* hb_features
     cdef bytes packed
     cdef char* cstr
     cdef hb_feature_t feat
     cdef const char **c_shapers
-    if features is None:
+    if not features:
         size = 0
         hb_features = NULL
     else:
-        size = len(features)
+        size = 0
+        for name, value in features.items():
+            if isinstance(value, int):
+                size += 1
+            else:
+                size += len(value)
         hb_features = <hb_feature_t*>malloc(size * sizeof(hb_feature_t))
-        for i, (name, value) in enumerate(features.items()):
+        i = 0
+        for name, value in features.items():
+            assert i < size, "index out of range for feature array capacity"
             packed = name.encode()
             cstr = packed
             hb_feature_from_string(packed, len(packed), &feat)
-            feat.value = value
-            hb_features[i] = feat
+            if isinstance(value, int):
+                feat.value = value
+                hb_features[i] = feat
+                i += 1
+            else:
+                for start, end, value in value:
+                    feat.value = value
+                    feat.start = start
+                    feat.end = end
+                    hb_features[i] = feat
```

Comment:
Hmm, is this re-use of feat really “safe”, shouldn’t you be declaring it inside the loop, or otherwise making a copy (this C but using Python syntax really makes me nervous). Also in this branch `hb_feature_from_string()` is superfous and we probably don’t want to allow the extended syntax for the tag her, so `feat.tag = hb_tag_from_string()` might be better.

Reply:
Yes it is safe, it really is C semantics with Python syntax. The `hb_features[i] = feat` really copies the value into the array — it is not a pointer. (There is a single `hb_feature_t feat` variable that is used over and over again, this is not new with this PR.)

I agree that `hb_feature_from_string()` would be superfluous, but users may depend on this behavior, so it would not be a backwards compatible change.

Reply:
> Yes it is safe, it really is C semantics with Python syntax. The `hb_features[i] = feat` really copies the value into the array — it is not a pointer. (There is a single `hb_feature_t feat` variable that is used over and over again, this is not new with this PR.)

I see. Sorry for the noise.

> I agree that `hb_feature_from_string()` would be superfluous, but users may depend on this behavior, so it would not be a backwards compatible change.

I mean only for the tuple value case (there is no backward compatability here) since accepting `{"aalt[2:4]": (3, 0, 5)}` is rather weird.

Reply:
> I mean only for the tuple value case (there is no backward compatability here) since accepting {"aalt[2:4]": (3, 0, 5)} is rather weird.

Ah, of course you're right. What would be the best/easiest way to get a `hb_tag_t` from a tag string?

Reply:
`hb_tag_from_string()`

Reply:
(Sorry for such a basic question, I had a bit of a blackout, could have found it myself easily.)